packages/stryker/src/input/InputFileResolver.ts

```diff
@@ -50,34 +53,77 @@ export default class InputFileResolver {
+      return this.expand(this.filePatterns);
     } else {
       return this.resolveFilesUsingGit();
     }
   }
 
-  private resolveFilesUsingGit(): Promise<string[]> {
-    return childProcessAsPromised
-      .exec(
+  private resolveMutateFiles() {
+    return this.expand(this.mutatePatterns, !shallowEquals(this.mutatePatterns, new Config().mutate));
+
+    function shallowEquals<T>(arr1: ReadonlyArray<T>, arr2: ReadonlyArray<T>): boolean {
+      if (arr1.length !== arr2.length) {
+        return false;
+      } else {
+        for (let i = 0; i < arr1.length; i++) {
+          if (arr1[i] !== arr2[i]) {
+            return false;
+          }
+        }
+        return true;
+      }
+    }
+  }
+
+  /**
+   * Takes a list of globbing patterns and expands them into files.
+   * If a patterns starts with a `!`, it negates the pattern.
+   * @param patterns The patterns to expand into files
+   */
+  private async expand(patterns: ReadonlyArray<string>, logAboutUselessPatterns = true): Promise<string[]> {
+    const fileSet = new Set<string>();
+    for (const pattern of patterns) {
+      if (pattern.startsWith(IGNORE_PATTERN_CHARACTER)) {
+        const files = await this.expandPattern(pattern.substr(1), logAboutUselessPatterns);
+        files.forEach(fileName => fileSet.delete(fileName));
+      } else {
+        const files = await this.expandPattern(pattern, logAboutUselessPatterns);
+        files.forEach(fileName => fileSet.add(fileName));
+      }
+    }
+    return Array.from(fileSet);
+  }
+
+  private async expandPattern(globbingExpression: string, logAboutUselessPatterns: boolean): Promise<string[]> {
+    const fileNames = (await glob(globbingExpression)).map(relativeFile => path.resolve(relativeFile));
+    if (!fileNames.length && logAboutUselessPatterns) {
+      this.log.warn(
+        `Globbing expression "${globbingExpression}" did not result in any files.`
+      );
+    }
+    return fileNames;
+  }
+
+  private async resolveFilesUsingGit(): Promise<string[]> {
+    try {
+      const { stdout } = await childProcessAsPromised.exec(
         'git ls-files --others --exclude-standard --cached --exclude .stryker-tmp',
         { maxBuffer: 10 * 1000 * 1024 }
-      )
-      .then(({ stdout }: { stdout: string }) => stdout.toString())
-      .then((output: string) =>
-        output.split('\n').map(fileName => fileName.trim())
-      )
-      .then((fileNames: string[]) =>
-        fileNames
-          .filter(fileName => fileName)
-          .map(fileName => path.resolve(fileName))
-      )
-      .catch((error: Error) => {
-        throw new StrykerError(
-          `Cannot determine input files. Either specify a \`files\` array in your stryker configuration, or make sure "${process.cwd()}" is located inside a git repository`,
-          error
-        );
-      });
+      );
+      const fileNames = stdout.toString()
+        .split('\n')
+        .map(line => line.trim())
+        .filter(line => line)
```

Comment:
Can you explain why this is necessary? It looks superfluous, but knowing you it probably isn't.

Reply:
This better?

```ts
        .filter(line => line) // remove empty lines
```

It filters out empty lines, as empty string is a falsy value.

By the way, I'm touched to hear that you think I'm not a suspicious person 🎅

Reply:
Yes, absolutely better IMHO.